### PR TITLE
Add more metrics exposed as trend in the new metric system

### PR DESCRIPTION
**Pull request by morucci**
related #847 

This change finalize the migration of metrics in the new metric system.

- [x] Add complex metrics trend support
- [ ] Expose long description in metric/list proto
- [x] Make the histo polymorphic by take the 'a'
- [ ] Support the slice periods as the Trend option

### CODE REVIEW

File: haskell/src/Monocle/Api/Server.hs [hunk 1]
```diff
@@ -713,9 +713,13 @@ metricGet auth request = checkAuth auth response
     runMetricQuery tenant query getRequestMetric getRequestOptions = do
       case (getRequestMetric, getRequestOptions) of
         ("changes_created_count", Nothing) -> intResult Q.metricChangesCreatedCount
+        ("changes_created_count", Just (MetricPB.GetRequestOptionsTrend _)) -> histoResult Q.metricChangesCreatedCountHisto
```
[the remark this answers is not in the document]
Comment (morucci) [reply haskell/src/Monocle/Api/Server.hs]: I've proposed a refactor. See commit 9b417f59162861273d7cc79037cb5be2a1ac1c4b

File: haskell/src/Monocle/Backend/Queries.hs [hunk 2]
```diff
@@ -1057,11 +1054,15 @@ getSuggestions index = do
...
+
+data MetricDataType = Num | Trend
+
 data MetricInfo = MetricInfo
   { miMetricName :: Text,
     miName :: Text,
     miDesc :: Text,
-    miLongDesc :: Maybe Text
+    miLongDesc :: Maybe Text,
+    miType :: [MetricDataType]
```
[the remark this answers is not in the document]
Comment (morucci) [reply haskell/src/Monocle/Backend/Queries.hs]: How "These" would be better than the MetricType sumtype ?

File: protos/monocle/protob/metric.proto [hunk 3]
```diff
@@ -35,20 +35,30 @@ message GetRequest {
...
+  repeated HistoFloat histo = 1;
 }
 
 message GetResponse {
   oneof result {
     string error = 1;
     float float_value = 2;
     int32 int_value = 3;
-    HistoStat histo_value = 4;
+    HistoIntStat histo_int_value = 4;
```
Comment (TristanCacqueray) [inline protos/monocle/protob/metric.proto:61]: I guess we could remove an indirection by using:
```suggestion
    repeated HistoInt histo_int = 4;
```

**Review by TristanCacqueray — APPROVED** [review]
:+1:

File: protos/monocle/protob/metric.proto (hunk 3 above)
Comment (morucci) [reply protos/monocle/protob/metric.proto:61]: In a "oneof" clause "repeated" keyword if not authorized unfortunately.

File: haskell/src/Monocle/Backend/Queries.hs [hunk 4]
```diff
@@ -1088,64 +1049,213 @@ rangeFlavorToDesc = \case
     "Both, the event and change's creation date is matched in "
       <> "case of any date query filter."
 
+monoHisto :: forall m a. QueryMonad m => m a -> m (V.Vector (Histo a))
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Queries.hs:1052]: :+1:

File: haskell/src/Monocle/Backend/Queries.hs [hunk 5]
```diff
@@ -1066,11 +1004,34 @@ data MetricInfo = MetricInfo
...
   { metricInfo :: MetricInfo,
-    runMetric :: m a
+    runMetric :: m (Numeric a),
+    runMetricTrend :: m (V.Vector (Histo a))
   }
 
-instance Functor m => Functor (Metric m) where
-  fmap f x = Metric (metricInfo x) (f <$> runMetric x)
+instance (Functor m) => Functor (Metric m) where
+  fmap f x = Metric (metricInfo x) (convN <$> runMetric x) (convT <$> runMetricTrend x)
```
Comment (TristanCacqueray) [inline haskell/src/Monocle/Backend/Queries.hs:1012]: If Numeric and Histo implement Functor, then we could remove the helper like this:

```suggestion
  fmap f x = Metric (metricInfo x) (fmap f <$> runMetric x) (fmap f <$> runMetricTrend x)
```

And using pattern match:

```suggestion
  fmap f Metric {..} = Metric metricInfo (fmap f <$> runMetric) (fmap f <$> runMetricTrend)
```

**Review by TristanCacqueray — COMMENTED** [review]
:+1:

**Pull request merged**